proof/refine/ARM/Finalise_R.thy

```diff
@@ -2640,6 +2640,97 @@ crunches cancelSignal, cancelAllIPC
+  "setSchedContext p sc \<lbrace> untyped_ranges_zero' \<rbrace>"
+  by (wp untyped_ranges_zero_lift)
+
+lemma threadSet_valid_queues_no_state:
+  "\<lbrace>Invariants_H.valid_queues and (\<lambda>s. \<forall>p. t \<notin> set (ksReadyQueues s p))\<rbrace>
+   threadSet f t
+   \<lbrace>\<lambda>_. Invariants_H.valid_queues\<rbrace>"
+  apply (simp add: threadSet_def)
+  apply wp
+   apply (simp add: valid_queues_def valid_queues_no_bitmap_def' pred_tcb_at'_def)
+   apply (wp hoare_Ball_helper
+             hoare_vcg_all_lift
+             setObject_tcb_strongest)[1]
+  apply (wp getObject_tcb_wp)
+  apply (clarsimp simp: valid_queues_def valid_queues_no_bitmap_def' pred_tcb_at'_def)
+  apply (clarsimp simp: obj_at'_def)
+  done
+
+lemma threadSet_valid_queues'_no_state:
+  "(\<And>tcb. tcbQueued tcb = tcbQueued (f tcb)) \<Longrightarrow>
+   \<lbrace>valid_queues' and (\<lambda>s. \<forall>p. t \<notin> set (ksReadyQueues s p))\<rbrace>
+   threadSet f t
+   \<lbrace>\<lambda>_. valid_queues'\<rbrace>"
+  apply (simp add: valid_queues'_def threadSet_def obj_at'_real_def
+                split del: if_split)
+  apply (simp only: imp_conv_disj)
+  apply (wp hoare_vcg_all_lift hoare_vcg_disj_lift)
+     apply (wp setObject_ko_wp_at | simp add: objBits_simps')+
+    apply (wp getObject_tcb_wp updateObject_default_inv
+               | simp split del: if_split)+
+  apply (clarsimp simp: obj_at'_def ko_wp_at'_def projectKOs
+                        objBits_simps addToQs_def
+             split del: if_split cong: if_cong)
+  apply (fastforce simp: projectKOs inQ_def split: if_split_asm)
+  done
+
+lemma schedContextUnbindTCB_invs'_helper:
+  "\<lbrace>\<lambda>s. invs' s \<and> scPtr \<noteq> idle_sc_ptr
+                \<and> ko_at' sc scPtr s
+                \<and> scTCB sc = Some tcbPtr
+                \<and> bound_sc_tcb_at' ((=) (Some scPtr)) tcbPtr s
+                \<and> (\<forall>a b. tcbPtr \<notin> set (ksReadyQueues s (a, b)))\<rbrace>
+   do threadSet (tcbSchedContext_update (\<lambda>_. Nothing)) tcbPtr;
+      setSchedContext scPtr $ scTCB_update (\<lambda>_. Nothing) sc
+   od
+   \<lbrace>\<lambda>_. invs'\<rbrace>"
+  unfolding schedContextUnbindTCB_def invs'_def valid_state'_def
+  apply (wp threadSet_valid_queues_no_state threadSet_valid_queues'_no_state
+            threadSet_valid_release_queue threadSet_valid_release_queue'
+            threadSet_not_inQ threadSet_idle' threadSet_iflive' threadSet_ifunsafe'T
+            threadSet_valid_pspace'T threadSet_sch_actT_P[where P=False, simplified]
+            threadSet_ctes_ofT threadSet_ct_idle_or_in_cur_domain' threadSet_cur
+            threadSet_global_refsT irqs_masked_lift untyped_ranges_zero_lift
+            valid_irq_node_lift valid_irq_handlers_lift''
+         | (rule hoare_vcg_conj_lift, rule threadSet_wp)
+         | clarsimp simp: tcb_cte_cases_def cteCaps_of_def)+
+  apply (frule ko_at_valid_objs'_pre[where p=scPtr], clarsimp)
+  (* slow *)
```

Comment:
How slow? Once it gets to around 30s I like to give an idea of how long it takes, to help people work out if something has gone wrong and it's gotten stuck.